src/instancemanager.jl

```diff
@@ -0,0 +1,327 @@
+        end
+    end
+    return vindices
+end
+
+function MOI.canaddconstraint(m::InstanceManager, func::MOI.AbstractFunction, set::MOI.AbstractSet)
+    MOI.canaddconstraint(m.instance, func, set) || return false
+    if m.state == AttachedSolver && m.mode == Manual
+        MOI.canaddconstraint(m.solver, func, set) || return false
+    end
+    return true
+end
+
+function MOI.addconstraint!(m::InstanceManager, func::MOI.AbstractFunction, set::MOI.AbstractSet)
+    # The canaddconstraint checks should catch most issues, but if an
+    # addconstraint! call fails then the instance and the solver may no longer
+    # be in sync.
+    if m.mode == Automatic && !MOI.canaddconstraint(m.solver, func, set)
+        resetsolver!(m)
+    end
+    @assert MOI.canaddconstraint(m, func, set)
+    cindex = MOI.addconstraint!(m.instance, func, set)
+    if m.state == AttachedSolver
+        cindex_solver = MOI.addconstraint!(m.solver, func, set)
+        m.instancetosolvermap[cindex] = cindex_solver
+    end
+    return cindex
+end
+
+# TODO: addconstraints!, canmodifyconstraint, modifyconstraint!,
+# transformconstraint!, cantransformconstraint
+
+## InstanceManager get and set attributes
+
+function MOI.set!(m::InstanceManager, attr::MOI.AbstractInstanceAttribute, value)
+    # The canset checks should catch most issues, but if a set! call fails then
+    # the instance and the solver may no longer be in sync.
+    if m.mode == Automatic && !MOI.canset(m.solver, attr)
+        resetsolver!(m)
+    end
+    @assert MOI.canset(m.instance, attr)
+    if m.state == AttachedSolver
+        @assert MOI.canset(m.solver, attr)
+        MOI.set!(m.solver, attr, value)
+    end
+    MOI.set!(m.instance, attr, value)
+end
+
+function MOI.set!(m::InstanceManager, attr::Union{MOI.AbstractVariableAttribute,MOI.AbstractConstraintAttribute}, ref, value)
+    if m.mode == Automatic && !MOI.canset(m.solver, attr, typeof(ref))
+        resetsolver!(m)
+    end
+    @assert MOI.canset(m.instance, attr, ref)
+    if m.state == AttachedSolver
+        @assert MOI.canset(m.solver, attr, typeof(ref))
+        MOI.set!(m.solver, attr, ref, value)
+    end
+    MOI.set!(m.instance, attr, ref, value)
+end
+
+# TODO: Automatic mode is broken in the case that the user tries to set
+# an objective function of a type that's not supported by the solver.
+# Two possible solutions are:
+# 1. Let canset take a value argument.
+# 2. Be more precise about what types are allowed in each attribute
+# (https://github.com/JuliaOpt/MathOptInterface.jl/issues/31).
+function MOI.canset(m::InstanceManager, attr::MOI.AbstractInstanceAttribute)
+    MOI.canset(m.instance, attr) || return false
+    if m.state == AttachedSolver && m.mode == Manual
+        MOI.canset(m.solver, attr) || return false
+    end
+    return true
+end
+
+function MOI.canset(m::InstanceManager, attr::Union{MOI.AbstractVariableAttribute,MOI.AbstractConstraintAttribute}, idxtype)
```

Comment:
In MOI, we define `canset(::AbstractInstance, ::AnyAttribute, ::Type{<:Index})` here you are more specific only in the two first arguments, this might create ambiguity so I'd annotate the type of the third arg with `Type{<:MOI.Index}`